shell/run.ts

```diff
@@ -0,0 +1,181 @@
+import {
+  FullRunOptions,
+  FullUserRunOptions,
+  PipedProcessResult,
+  ProcessResult,
+  UserRunOptions,
+  UserTryOptions,
+} from './run_types.ts';
+import { trimTrailingNewline } from '../lang/strings.ts';
+
+interface PipedRunOptions extends FullRunOptions {
+  stdout: 'piped';
+  stderr: 'piped';
+}
+
+/**
+ * A command to be run by a process.
+ *
+ * Can either be an array of command + arguments, or a single string with a
+ * complete statement that will be forwarded to the shell.
+ */
+export type ProcessCommand = string | string[];
+
+/**
+ * Runs a process and accepts options for handling I/O.
+ *
+ * By default, this mimics the `$(command)` functionality in BASH, meaning that it pipes all output and returns it as a
+ * string.
+ *
+ * @param command The {@link ProcessCommand} to be run
+ * @param options The {@link FullUserRunOptions} governing the behavior of the process runner
+ * @returns A {@link string} containing the piped `stdout` from the process
+ */
+export async function $(
+  command: ProcessCommand,
+  options: FullUserRunOptions = {},
+): Promise<string> {
+  const result = await runPiped(buildCommandArray(command), options);
+  return result.stdout;
+}
+
+/**
+ * Runs a process using the generic API.
+ *
+ * @param command The {@link ProcessCommand} to be run
+ * @param options The {@link FullUserRunOptions} governing the behavior of the process runner
+ * @returns A {@link Promise<ProcessResult>} containing the result state when the process completes.
+ */
+$.run = runGeneric;
+
+/**
+ * Runs a process using the piped output API.
+ *
+ * Note: This is the same as using {@link $} on its own, except that it provides the full process result instead of
+ * just the string output.
+ *
+ * @param command The {@link ProcessCommand} to be run
+ * @param options The {@link FullUserRunOptions} governing the behavior of the process runner
+ * @returns A {@link Promise<PipedProcessResult>} containing the result state and output when the process completes
+ */
+$.piped = runPiped;
+
+/**
+ * Runs a process using the streamed output API.
+ *
+ * This means that output will be sent to the same `stdout`/`stderr` as the Deno script, so it will show up in along with
+ * console output.
+ *
+ * @param command The {@link ProcessCommand} to be run
+ * @param options The {@link FullUserRunOptions} governing the behavior of the process runner
+ * @returns A {@link Promise<ProcessResult>} containing the result state when the process completes
+ */
+$.streamed = runStreamed;
+
+/**
+ * Runs a process and checks to see if it was successful.
+ *
+ * @param command The {@link ProcessCommand} to be run
+ * @param options The {@link FullUserRunOptions} governing the behavior of the process runner
+ * @returns A {@link Promise<boolean>} that will resolve to `true` if the process had a zero exit code to indicate success
+ */
+$.try = tryRun;
+
+async function tryRun(
+  command: string[] | string,
```

Comment:
I would suggest being consistent by either using `ProcessCommand` throughout, or alternatively by removing the `ProcessCommand` type and just using this literal type. It was a little confusing to me that the documentation indicates that `command` is a `ProcessCommand` but here in the implementation it is expressed as the underlying `string | string[]`.

Reply:
Good catch! That was super confusing.